### PR TITLE
refactor: add ForecastBatch contract

### LINKED ISSUE #613: [Enhancement] Stabilize the WeatherDataset batch contract for boundary/global extensions
### Current problem

`WeatherDataset.__getitem__` currently returns a positional 4-tuple:

- `init_states`
- `target_states`
- `forcing`
- `target_times`

and downstream code relies on positional access / unpacking.

Concrete examples in the current codebase:
- `neural_lam/weather_dataset.py` returns the raw tuple in `__getitem__`
- `neural_lam/datastore/npyfilesmeps/compute_standardization_stats.py` destructures the batch directly in the loader loop
- current tests also encode the tuple contract directly

This is workable for the current deterministic path, but it becomes increasingly brittle as we add or vary:
- optional boundary forcing
- time-delta / time-encoding features
- no-boundary / global edge cases
- more general multi-datastore inputs later

The issue is not performance, but contract clarity: positional accesses make semantics harder to follow, easy to break silently, and awkward to extend with optional fields.

### Why this seems useful now

This looks like a good infrastructural step because it sits directly between two active directions already present in the repo:

- #138 is pushing toward boundary/global handling where the data contract becomes more complex
- #335 is discussing the forecast/output contract for deterministic vs probabilistic models

In that sense, this issue is the data-side counterpart to #335:
- #335 asks what forecast tensors mean once models become probabilistic
- this issue asks what a training/inference batch means before it reaches the model

A clearer batch contract should make boundary/global extensions easier to add without forcing those features to keep depending on fragile tuple ordering.

### Proposed enhancement

Introduce a named batch structure, e.g. `ForecastBatch`, used as the return type of `WeatherDataset` and as the main interface between dataset and model code.

The intended implementation is still narrow in concept, but end-to-end in scope:
- define a lightweight batch container (dataclass / mapping / typed object)
- centralize batch unpacking / validation
- migrate the current 4 fields to named access
- update the current downstream consumers so the positional tuple contract is removed from the main deterministic training/evaluation path

The batch would initially name fields like:
- `init_states`
- `target_states`
- `forcing`
- `target_times`

and should be able to grow cleanly with optional fields such as:
- `boundary_forcing`
- `future_boundary_forcing`
- `time_deltas`

without requiring every downstream consumer to be rewritten around tuple-order assumptions.

### Intended PR scope

The intended implementation is one coherent infrastructure PR that updates the current batch contract end-to-end across the existing deterministic path, including:
- `WeatherDataset.__getitem__`
- batch unpacking / validation helpers
- `compute_standardization_stats.py`
- tests covering the named batch contract and affected paths

The goal is to land the contract migration as one reviewable infrastructure change, rather than several partial transition PRs that preserve pieces of the positional interface.

### Non-goals

This is **not** proposing:
- a broad redesign of the whole training stack
- multi-datastore composition in full generality
- stochastic rollout or diffusion sampling interfaces
- probabilistic metrics / CRPS / spread-skill work
- graph or connectivity changes
- ensemble output contracts (those belong more naturally with #335)

The aim is only to replace the fragile positional batch contract with a named one, in a way that remains compatible with current functionality and supports later boundary/global extensions more cleanly.

### Related

Refs #138  
Complements #335

### PR DESCRIPTION
**Pull request by kshirajahere**
Refs #613

## Describe your changes

Adds a named `ForecastBatch` contract for `WeatherDataset` outputs and migrates the current post-#208 internal consumers away from positional batch access.

## Changes
- Add `ForecastBatch(NamedTuple)` in `neural_lam/batch.py`.
- Return `ForecastBatch` from `WeatherDataset.__getitem__`.
- Update `ForecasterModule`, `compute_standardization_stats.py`, and tests to use named fields.
- Keep native PyTorch `default_collate` support without a custom collate function.
- Add regression coverage for DataLoader collation and named batch access.

## Issue Link
Closes #613 

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [x] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [x] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [x] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by kshirajahere** [comment]
@sadamov @leifdenby please have a look 🙏 thank you

File: neural_lam/batch.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    target_times: torch.Tensor
+
+    def to(self, device: torch.device | str) -> "ForecastBatch":
+        """Move all tensor fields to the same device."""
+        return ForecastBatch(
+            init_states=self.init_states.to(device),
+            target_states=self.target_states.to(device),
+            forcing=self.forcing.to(device),
+            target_times=self.target_times.to(device),
+        )
```
Comment (sadamov) [inline neural_lam/batch.py:31]: Add `boundary` and propagate through `.to()`:

```suggestion
    init_states: torch.Tensor
    target_states: torch.Tensor
    forcing: torch.Tensor
    boundary: torch.Tensor
    target_times: torch.Tensor

    def to(self, device: torch.device | str) -> "ForecastBatch":
        """Move all tensor fields to the same device."""
        return ForecastBatch(
            init_states=self.init_states.to(device),
            target_states=self.target_states.to(device),
            forcing=self.forcing.to(device),
            boundary=self.boundary.to(device),
            target_times=self.target_times.to(device),
        )
```

File: neural_lam/batch.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            "Expected a ForecastBatch or 4-element legacy batch tuple, "
+            f"got {len(batch)} elements."
+        )
+
+    return ForecastBatch(
+        init_states=batch[0],
+        target_states=batch[1],
+        forcing=batch[2],
+        target_times=batch[3],
+    )
```
Comment (sadamov) [inline neural_lam/batch.py:56]:
```suggestion
    return ForecastBatch(
        init_states=batch[0],
        target_states=batch[1],
        forcing=batch[2],
        boundary=batch[3],
        target_times=batch[4],
    )
```

File: neural_lam/weather_dataset.py [hunk 3]
```diff
@@ -527,7 +527,12 @@ def __getitem__(self, idx):
         # forcing: (ar_steps, N_grid, d_windowed_forcing)
         # target_times: (ar_steps,)
 
-        return init_states, target_states, forcing, target_times
+        return ForecastBatch(
+            init_states=init_states,
+            target_states=target_states,
+            forcing=forcing,
+            target_times=target_times,
+        )
```
Comment (sadamov) [inline neural_lam/weather_dataset.py:531]: Pass `boundary` through here too (zero tensor of matching shape is fine until #635 wires up real loading).

File: neural_lam/models/module.py [hunk 4]
```diff
@@ -176,13 +177,14 @@ def configure_optimizers(self):
...
+        batch = coerce_forecast_batch(batch)
         prediction, pred_std = self.forecaster(
-            init_states, forcing_features, target_states
+            batch.init_states, batch.forcing, batch.target_states
         )
-        return prediction, target_states, pred_std, batch_times
+        return prediction, batch.target_states, pred_std, batch.target_times
 
     def training_step(self, batch):
+        batch = coerce_forecast_batch(batch)
```
Comment (sadamov) [inline neural_lam/models/module.py:384]: Nit: `training_step` / `validation_step` / `test_step` each coerce, then `common_step` coerces again. Could lift to one place. Not blocking.

File: CHANGELOG.md [hunk 5]
```diff
@@ -35,6 +35,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Maintenance
 
+- Add a named `ForecastBatch` contract for `WeatherDataset` batches and update
+  internal consumers to use field access [\#613](https://github.com/mllam/neural-lam/issues/613) @kshirajahere
```
Comment (sadamov) [inline CHANGELOG.md]: Link the PR, not the issue.

```suggestion
- Add a named `ForecastBatch` contract for `WeatherDataset` batches and update
  internal consumers to use field access [\#637](https://github.com/mllam/neural-lam/pull/637) @kshirajahere
```

**Review by sadamov — COMMENTED** [review]
Nice PR @kshirajahere, thanks! One ask before merge: please add `boundary` as a 4th field of `ForecastBatch` now, so this lands cleanly alongside #635 in v0.8.0 without a re-rebase of every consumer. Relabelled as enhancement, proposed for v0.8.0.

**Comment by leifdenby** [comment]
I really appreciate the effort to construct a data-structure for the entire training sample, rather than just a tuple of tensors, this will be fantastic 🌟 . But I think having a key in the named-tuple called `boundary` as well as keys with names `state`, `static`, `forcing` doesn't quite make sense relative to the discussions we've been having on #635. This is because the `state,static,forcing` labels denote what these tensors are used for (i.e. are they input/output, input but needs broadcasting across the batch and time, or only input) whereas `boundary` denotes *what spatial domain does the data come from*. I think it is a bit confusing to mix the two. Or said different, from just the name `boundary` it isn't obvious to me if this boundary-forcing data or boundary-static data.

What we probably need is either:
1. Data-structure where both spatial domain and use of data is explicit through a nested structure, i.e. `batch.{domain}.{data_category}`, which means there would be tensors like `batch.interior.state` and `batch.boundary.forcing`, or
2. We coerce the spatial domain and data category into a single key, i.e. `batch.{domain}__{data_category}` with examples like `batch.interior__state` and `batch.boundary__forcing`

I would prefer 1. above, coercing multiple things into a single string is always brittle. In addition, I would actually advocate for the `{domain}` identifier above we actuallyuse the datastore names that are being introduced https://github.com/mllam/neural-lam/pull/635 (based on discussions in #652), so that the data-structure becomes `batch.{datastore_name}.{data_category}` with examples `batch.danra.state` and `batch.era5.forcing`.

Thoughts? :)

**Comment by joeloskarsson** [comment]
This was planned for version 0.8.0, conditioned on some refining of the exact names of the different tensors in the Batch construct.
